File: C/chapitre7/Exercice3.c
#include <stdio.h>
#include <unistd.h>
#include <signal.h>
#include <sys/wait.h>

void handler(int s) {
    printf("Fils %d tué\n", getpid());
    _exit(0);
}

pid_t fils[5];

int main(void) {
    int i;
    for (i = 0; i < 5; i++) {
        fils[i] = fork();
        if (fils[i] == 0) {
            signal(SIGTERM, handler);
            while (1) { }
        }
    }
    char cmd;
    int num;
    while (scanf(" %c %d", &cmd, &num) == 2) {
        if (cmd == 's') kill(fils[num], SIGSTOP);
        else if (cmd == 'r') kill(fils[num], SIGCONT);
        else if (cmd == 't') kill(fils[num], SIGTERM);
    }
    return 0;
}

File: Python/chapitre7/Exercice3.py
import os
import signal
import sys
def handler(s, f):
    print(f"Fils {os.getpid()} tué")
    os._exit(0)
fils = []
for i in range(5):
    p = os.fork()
    if p == 0:
        signal.signal(signal.SIGTERM, handler)
        while True:
            pass
    fils.append(p)
while True:
    line = sys.stdin.readline().split()
    if not line:
        break
    cmd = line[0]
    num = int(line[1])
    if cmd == 's':
        os.kill(fils[num], signal.SIGSTOP)
    elif cmd == 'r':
        os.kill(fils[num], signal.SIGCONT)
    elif cmd == 't':
        os.kill(fils[num], signal.SIGTERM)